Various fixes to C# warnings

raulsntos opened the pull request:
- Suppress CS0108 caused by shadowing members in derived classes.
	- Fixes https://github.com/godotengine/godot/issues/15763
- Suppress CA1716 caused by parameter names matching C# keywords.
	- Follow up to https://github.com/godotengine/godot/pull/87518.

<details>
<summary><h3>Extracted to different PRs</h3></summary>

- Suppress REFL045 in a false positive scenario.
	- Extracted to https://github.com/godotengine/godot/pull/88570.
- Suppress CA1710 caused by Array not having the Collection suffix.
	- Extracted to https://github.com/godotengine/godot/pull/87518.
- Suppress CA1716 caused by parameter names matching C# keywords.
	- Extracted to https://github.com/godotengine/godot/pull/87518.
- Use SymbolEqualityComparer in ScriptPathAttributeGenerator to fix RS1024.
	- Extracted to https://github.com/godotengine/godot/pull/87342.
- Suppress dependencies of SourceGenerators package to fix NU5128.
	- Extracted to https://github.com/godotengine/godot/pull/79501.
- Suppress nullability warnings in source generators because we already have a Debug.Assert (but it's missing the nullability annotations in netstandard2.0).
	- Extracted to https://github.com/godotengine/godot/pull/87518.
- Implement disposable pattern in `GodotSynchronizationContext` (and make it sealed).
	- Extracted to https://github.com/godotengine/godot/pull/72053.
- Fix reference to `NotificationSceneInstantiated` in the documentation.
	- Follow up to https://github.com/godotengine/godot/pull/64410.
	- Extracted to https://github.com/godotengine/godot/pull/72462.
- Fix RS2008: Enable analyzer release tracking.
	- Implemented by https://github.com/godotengine/godot/pull/80343.

</details>

### TODO
- [ ] Fix CA1001: Type owns disposable field but is not disposable.
	- [ ] `Array` and `Dictionary`: This was already discussed a bit in https://github.com/godotengine/godot/pull/64089#discussion_r940589373 but I think it's not as straightforward as the rest of the fixes and it should probably be handled in a separate PR.
	- [ ] `Variant`: It already implements `IDisposable` so I don't know why it complains.
		- Seems like it's a bug in the analyzer: https://github.com/dotnet/roslyn-analyzers/issues/6151.

neikeq commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/ScriptMethodsGenerator.cs:
```diff
@@ -135,7 +136,9 @@ INamedTypeSymbol symbol
             source.Append("#pragma warning disable CS0109 // Disable warning about redundant 'new' keyword\n");
 
-            source.Append($"    public new class MethodName : {symbol.BaseType.FullQualifiedName()}.MethodName {{\n");
+            Debug.Assert(symbol.BaseType != null, "Symbol always has a base type because every script class must inherit from a Godot type; otherwise, the generator wouldn't be visiting this class.");
```
I really don't like the fact flow analysis can't understand the value cannot be null after this check in .NET Standard 2.0. I think throwing an exception would be better.

neikeq replied:
Actually, wouldn't this check be excluded when publishing the package?

```cs
[System.Diagnostics.Conditional("DEBUG")]
public static void Assert (bool condition);
```

Calls to this function are only compiled in Debug configurations. It's the same as:

```cs
#if DEBUG
Debug.Assert(foo);
#endif
```

raulsntos replied:
Yes, the call is excluded, that's why I'm using it in these cases because they will never be null so I wanted to avoid the unnecessary check but it might still be useful during development to notice if we're doing something wrong.

neikeq commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -953,7 +953,9 @@ void BindingsGenerator::_generate_global_constants(StringBuilder &p_output) {
-					"'Missing XML comment for publicly visible type or member'\n");
+					"'Missing XML comment for publicly visible type or member'\n"
+					"#pragma warning disable CA1716 // Disable warning: "
+					"'Identifiers should not match keywords'\n");
```
I don't know how to feel about this. I would like to keep this in sight. We may want to rename those in the future. The Object one if fine to disable, because it's very obvious.

neikeq replied:
Can this be done inside an `#if`? If so, we could have a configuration to enable this warning, while having it disabled by default.

raulsntos replied:
It can and it sounds good to me. I don't really want to suppress these warnings but solving them properly would require renaming which seems unlikely to happen in a timely manner and after the beta we won't want to break compatibility.

neikeq commented on modules/mono/editor/bindings_generator.cpp:
```diff
@@ -1370,10 +1373,14 @@ Error BindingsGenerator::_generate_cs_type(const TypeInterface &itype, const Str
-				  "#pragma warning disable CS1573 // Disable warning: "
-				  "'Parameter has no matching param tag in the XML comment'\n");
+				  "#pragma warning disable CA1716 // Disable warning: "
+				  "'Identifiers should not match keywords'\n");
```
Ditto. Can `CS0108` and `CA1716` be done inside an `#if`?

neikeq commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs:
```diff
@@ -5,6 +5,9 @@
 using Godot.NativeInterop;
 
+// Suppress warning because it complaints about Array not having the Collection suffix.
+#pragma warning disable CA1710 // Identifiers should have correct suffix
```
Really nonsense warning. We may want to disable it directly in `.editorconfig`.

neikeq commented on modules/mono/glue/GodotSharp/GodotSharp/Core/Extensions/ObjectExtensions.cs:
```diff
@@ -1,6 +1,8 @@
 using System;
 using Godot.NativeInterop;
 
+#pragma warning disable CA1716 // Disable warning: 'Identifiers should not match keywords'
```
Ditto. Can this be done inside an `#if`?

neikeq commented on modules/mono/glue/GodotSharp/GodotSharp/Core/GodotSynchronizationContext.cs:
```diff
@@ -24,5 +24,10 @@ public void ExecutePendingContinuations()
+        public void Dispose()
+        {
+            _queue.Dispose();
+        }
```
Hmm maybe we should dispose this somewhere.

raulsntos replied:
Interesting, it's used in `GodotTaskScheduler` but I don't remember getting the same warning about it not being disposable there, I guess it should also be made disposable.

These types seem to be mostly used as singletons (they are static fields) so can't really dispose them. I can dispose `GodotTaskScheduler` in `Dispatcher.InitializeDefaultGodotTaskScheduler` if the static instance wasn't null before assigning although I expect this method to only be called once.

neikeq commented:
> Fix RS2008: Enable analyzer release tracking.
> 
>     * Not sure how to fix this one. The quick fix in VSCode does nothing.

I think this is about https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md

neikeq commented:
> Fix CA1001: Type owns disposable field but is not disposable.
> 
>     * [ ]  `Array` and `Dictionary`: This was already discussed a bit in

This won't be a simple change. It will require some design decisions.

raulsntos commented on modules/mono/editor/Godot.NET.Sdk/Godot.SourceGenerators/Godot.SourceGenerators.csproj:
```diff
@@ -18,6 +18,7 @@
     <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
     <!-- Do not include the generator as a lib dependency -->
     <IncludeBuildOutput>false</IncludeBuildOutput>
+    <SuppressDependenciesWhenPacking>true</SuppressDependenciesWhenPacking>
```
The warning I was getting is [Scenario 2](https://docs.microsoft.com/en-us/nuget/reference/errors-and-warnings/nu5128#scenario-2). Since this package is a source generator it does not contain anything in `lib/` or `ref/` (only in `analyzers/`) because it's not a library and this warning is complaining that there's no assembly in `lib/` for the TFM `netstandard2.0` that we target.

So I think this situation is solved using this one of the solutions described in the documentation:

> If your package does not contain any `lib/` or `ref/` files and is not a meta-package, it likely does not have any dependencies that the package consumer needs. If you are packing with NuGet's MSBuild Pack target, you can set `<SuppressDependenciesWhenPacking>true</SuppressDependenciesWhenPacking>` in any `PropertyGroup` in your project file. If you are using a custom `nuspec` file, remove the `<dependencies>` element.

I don't fully understand though, so maybe I'm wrong.

raulsntos commented:
> > Fix RS2008: Enable analyzer release tracking.
> > ```
> > * Not sure how to fix this one. The quick fix in VSCode does nothing.
> > ```
> 
> I think this is about [https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md](https://github.com/dotnet/roslyn-analyzers/blob/main/src/Microsoft.CodeAnalysis.Analyzers/ReleaseTrackingAnalyzers.Help.md?rgh-link-date=2022-09-09T20%3A18%3A03Z)

Yes but since the quick fix doesn't seem to do anything in VSCode, should I create the markdown files manually? Suppress the warning?

> > Fix CA1001: Type owns disposable field but is not disposable.
> > ```
> > * [ ]  `Array` and `Dictionary`: This was already discussed a bit in
> > ```
> 
> This won't be a simple change. It will require some design decisions.

I don't intend to solve this warning in this PR but I wanted to mention it since it's somewhat related.

raulsntos commented:
Rebased and extracted breaking changes to https://github.com/godotengine/godot/pull/72053.

raulsntos committed:
Various fixes to C# warnings

- Suppress REFL045 in a false positive scenario.
- Suppress CS0108 caused by shadowing members in derived classes.
- Move CA1716 to pragmas in order to allow re-enabling it with a custom define.

paulloz commented:
Can we close this one? The only remaining point is the several instances of `CA1001`, and this PR does not fix them. We should track that particular issue in its own thread instead.

raulsntos commented:
Yes, this has been superseded by the multiple PRs that are now linked in the PR description.
